File: scripts/verify_guest.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Run through a real dev SSH session. Never reads real credential contents.
require 'digest'
require 'etc'
require 'tmpdir'
require_relative 'check_codex'

module GuestVerification
  def self.check
    Verification.assert(Etc.getpwuid.name == 'dev', 'Must run as dev')
    Verification.assert(!ENV.key?('SSH_AUTH_SOCK'), 'Agent socket was forwarded')
    Verification.assert(!Verification.run(['sudo', '-n', 'true']).last.success?, 'dev has sudo access')
    Verification.assert((Process.groups - [Etc.getpwnam('dev').gid]).empty?, 'Unexpected supplementary groups')
    %w[/home/jack /root].each do |path|
      begin
        Dir.children(path)
      rescue Errno::EACCES
        next
      end
      raise "Administrator home accessible: #{path}"
    end
    %w[/etc/codex /etc/codex/requirements.toml /usr/local/bin /usr/local/share/agent-vm].each do |path|
      Verification.assert(File.stat(path).uid.zero? && !File.writable?(path), "Writable policy/tool path: #{path}")
    end
    mounts = File.read('/proc/mounts')
    Verification.assert(%w[virtiofs 9p fuse.sshfs].none? { |type| mounts.include?(type) }, 'Unexpected shared filesystem')
    Verification.assert(!File.exist?('/run/host-services/ssh-auth.sock'), 'Host agent socket is present')
    Verification.assert(!(File.readable?('/var/run/docker.sock') && File.writable?('/var/run/docker.sock')), 'Docker socket accessible')
    # Compare the exact provisioned policy, without adding a TOML parser dependency.
    # check_codex validates that Codex actually loads its managed profile/features.
    expected = File.read('/usr/local/share/agent-vm/requirements.sha256').split.first
    Verification.assert(Digest::SHA256.file('/etc/codex/requirements.toml').hexdigest == expected, 'Managed policy differs from provisioned recipe')
    version = File.read('/usr/local/share/agent-vm/codex-version').strip
    output, error, status = Verification.run(['codex', '--version'])
    Verification.assert(status.success? && output.strip == "codex-cli #{version}", error)
    puts 'PASS Linux account, policy ownership, mounts, SSH forwarding, and Codex installation'
    CodexPolicyCheck.check
    sandbox_check
    puts 'NOT TESTED: authenticated model run, private GitHub repository scope, desktop-provided tool inventory'
  end

  def self.sandbox_check
    canary = File.join(Dir.home, '.pgpass')
    if File.exist?(canary) || File.symlink?(canary)
      raise 'Cannot run canary test: ~/.pgpass already exists; no secrets were read'
    end
    Dir.mktmpdir('verify-', File.join(Dir.home, 'projects')) do |fixture|
      work = File.join(fixture, 'workspace')
      Dir.mkdir(work)
      sibling = File.join(fixture, 'outside-workspace')
      File.write(sibling, 'original')
      created = false
      begin
        File.open(canary, File::WRONLY | File::CREAT | File::EXCL, 0o600) do |file|
          created = true
          file.write("agent-vm-synthetic-canary\n")
        end
        Verification.assert(File.read(canary) == "agent-vm-synthetic-canary\n", 'Canary unavailable outside sandbox')
        code = <<~'CHECK'
          File.write('workspace-write-ok', 'ok')
          begin
            File.binread(File.join(Dir.home, '.pgpass'))
            abort 'managed deny-read did not hold'
          rescue Errno::EACCES, Errno::EPERM
          end
          begin
            File.write(ARGV.fetch(0), 'changed')
            abort 'outside-workspace write succeeded'
          rescue Errno::EACCES, Errno::EPERM, Errno::EROFS
          end
          puts 'PASS Codex workspace write, outside-workspace write denial, and managed secret read denial'
        CHECK
        [[], ['-c', 'permissions.vm_dev.filesystem."~/.pgpass"="read"']].each do |overrides|
          output, error, status = Verification.run(['codex', 'sandbox', '--include-managed-config', '-P', 'vm_dev', '-C', work,
                                                   *overrides, 'ruby', '-e', code, sibling])
          if overrides.empty?
            Verification.assert(status.success? && output.include?('PASS Codex'), error + output)
            puts output.strip
          else
            Verification.assert(!status.success? && error.include?('conflicts with a config-defined profile'), error)
          end
        end
        Verification.assert(File.read(sibling) == 'original', 'Outside-workspace file changed')
        puts 'PASS conflicting config override rejected before execution'
      ensure
        File.unlink(canary) if created
      end
    end
  end
end

if $PROGRAM_NAME == __FILE__
  begin
    GuestVerification.check
  rescue StandardError => e
    warn e.message
    exit 1
  end
end

File: scripts/verify_guest.py
#!/usr/bin/env python3
"""Credential-free Linux and Codex checks, run through a real dev SSH session."""
import errno
import json
import os
from pathlib import Path
import pwd
import subprocess
import tempfile
import tomllib


def run(args, **kwargs):
    return subprocess.run(args, text=True, capture_output=True, timeout=45, **kwargs)


def main():
    assert pwd.getpwuid(os.getuid()).pw_name == "dev", "Must run as dev"
    assert os.environ.get("SSH_AUTH_SOCK") is None, "Agent socket was forwarded"
    assert run(["sudo", "-n", "true"]).returncode != 0, "dev has sudo access"
    assert set(os.getgroups()) <= {pwd.getpwnam("dev").pw_gid}, "Unexpected supplementary groups"
    for path in ("/home/jack", "/root"):
        try:
            os.listdir(path)
        except PermissionError:
            pass
        else:
            raise AssertionError(f"Administrator home accessible: {path}")
    for path in ("/etc/codex", "/etc/codex/requirements.toml", "/usr/local/bin", "/usr/local/share/agent-vm"):
        assert os.stat(path).st_uid == 0 and not os.access(path, os.W_OK), f"Writable policy/tool path: {path}"
    mounts = Path("/proc/mounts").read_text()
    assert not any(t in mounts for t in ("virtiofs", "9p", "fuse.sshfs")), "Unexpected shared filesystem"
    assert not Path("/run/host-services/ssh-auth.sock").exists()
    assert not os.access("/var/run/docker.sock", os.R_OK | os.W_OK)
    req = tomllib.loads(Path("/etc/codex/requirements.toml").read_text())
    assert req["allowed_permission_profiles"] == {"vm_dev": True}
    assert req["mcp_servers"] == {}
    assert all(req["features"][k] is False for k in ("apps", "plugins", "browser_use", "in_app_browser", "computer_use"))
    version = Path("/usr/local/share/agent-vm/codex-version").read_text().strip()
    assert run(["codex", "--version"]).stdout.strip() == "codex-cli " + version
    print("PASS Linux account, policy ownership, mounts, SSH forwarding, and Codex installation")
    result = run(["python3", "/usr/local/share/agent-vm/check_codex.py"])
    assert result.returncode == 0, result.stderr
    print(result.stdout.strip())

    # Use a synthetic denied file only; never read or overwrite an existing secret.
    canary = Path.home() / ".pgpass"
    if canary.exists() or canary.is_symlink():
        raise RuntimeError("Cannot run canary test: ~/.pgpass already exists; no secrets were read")
    with tempfile.TemporaryDirectory(prefix="verify-", dir=Path.home() / "projects") as fixture:
        work = str(Path(fixture) / "workspace")
        Path(work).mkdir()
        sibling = Path(fixture) / "outside-workspace"
        sibling.write_text("original")
        created = False
        try:
            with canary.open("x") as f:
                created = True
                f.write("agent-vm-synthetic-canary\n")
            # Confirm it exists and is readable outside the sandbox.
            assert canary.read_text() == "agent-vm-synthetic-canary\n"
            code = '''import errno, pathlib, sys
pathlib.Path("workspace-write-ok").write_text("ok")
try:
    pathlib.Path.home().joinpath(".pgpass").read_bytes()
except OSError as exc:
    assert exc.errno in (errno.EACCES, errno.EPERM), repr(exc)
else:
    raise AssertionError("managed deny-read did not hold")
try:
    pathlib.Path(sys.argv[1]).write_text("changed")
except OSError as exc:
    assert exc.errno in (errno.EACCES, errno.EPERM, errno.EROFS), repr(exc)
else:
    raise AssertionError("outside-workspace write succeeded")
print("PASS Codex workspace write, outside-workspace write denial, and managed secret read denial")
'''
            for overrides in ([], ["-c", 'permissions.vm_dev.filesystem."~/.pgpass"="read"']):
                result = run(["codex", "sandbox", "--include-managed-config", "-P", "vm_dev", "-C", work,
                              *overrides, "python3", "-c", code, str(sibling)])
                if overrides:
                    assert result.returncode != 0 and "conflicts with a config-defined profile" in result.stderr, result.stderr
                    continue
                assert result.returncode == 0, result.stderr + result.stdout
                assert "PASS Codex" in result.stdout, result.stdout
                print(result.stdout.strip())
            assert sibling.read_text() == "original"
            print("PASS conflicting config override rejected before execution")
        finally:
            if created:
                canary.unlink()
    print("NOT TESTED: authenticated model run, private GitHub repository scope, desktop-provided tool inventory")


if __name__ == "__main__":
    main()
